src/common/mpi.jl

```diff
@@ -7,11 +7,25 @@ Number of processors used in MPI. Can be called without ensuring initialization.
 mpi_nprocs(comm=MPI.COMM_WORLD) = (MPI.Init(); MPI.Comm_size(comm))
 mpi_master(comm=MPI.COMM_WORLD) = (MPI.Init(); MPI.Comm_rank(comm) == 0)
 
-mpi_sum(  arr, comm::MPI.Comm) = MPI.Allreduce( arr,   +, comm)
-mpi_sum!( arr, comm::MPI.Comm) = MPI.Allreduce!(arr,   +, comm)
-mpi_min(  arr, comm::MPI.Comm) = MPI.Allreduce( arr, min, comm)
-mpi_min!( arr, comm::MPI.Comm) = MPI.Allreduce!(arr, min, comm)
-mpi_max(  arr, comm::MPI.Comm) = MPI.Allreduce( arr, max, comm)
-mpi_max!( arr, comm::MPI.Comm) = MPI.Allreduce!(arr, max, comm)
-mpi_mean( arr, comm::MPI.Comm) = mpi_sum(arr, comm) ./ mpi_nprocs(comm)
-mpi_mean!(arr, comm::MPI.Comm) = (mpi_sum!(arr, comm); arr ./= mpi_nprocs(comm))
+@static if Base.Sys.ARCH == :aarch64
+    # Custom reduction operators are not supported on aarch64 (see
+    # https://github.com/JuliaParallel/MPI.jl/issues/404), so we define fallback no-op
+    # mpi_* functions to get things working while waiting for an upstream solution.
+    mpi_sum(  arr, ::MPI.Comm) = arr
+    mpi_sum!( arr, ::MPI.Comm) = arr
+    mpi_min(  arr, ::MPI.Comm) = arr
+    mpi_min!( arr, ::MPI.Comm) = arr
+    mpi_max(  arr, ::MPI.Comm) = arr
+    mpi_max!( arr, ::MPI.Comm) = arr
+    mpi_mean( arr, ::MPI.Comm) = arr
+    mpi_mean!(arr, ::MPI.Comm) = arr
```

Comment:
Just FYI (not necessarily to do it here), a nice metaprogramming pattern here is `for fun in (:mpi_sum, :mpi_sum!, ...) @eval $fun(x) = x`

Reply:
Also, just stylewise, it's maybe better to put the common case first, and the workaround later, ie `if ARCH != :aarch64`